indexer.go

```diff
@@ -0,0 +1,863 @@
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+
+	car2 "github.com/ipld/go-car/v2"
+	selectorparse "github.com/ipld/go-ipld-prime/traversal/selector/parse"
+
+	ipld "github.com/ipld/go-ipld-prime"
+
+	dagcbor "github.com/ipld/go-ipld-prime/codec/cbor"
+	dagjson "github.com/ipld/go-ipld-prime/codec/json"
+
+	datamodel "github.com/ipld/go-ipld-prime/datamodel"
+	qp "github.com/ipld/go-ipld-prime/fluent/qp"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basicnode"
+
+	tree "github.com/kenlabs/go-ipld-prolly-trees/pkg/tree"
+)
+
+type Database struct {
+	blockStore  *blockstore.Blockstore
+	nodeStore   *tree.BlockNodeStore
+	rootCid     cid.Cid
+	tree        *tree.ProllyTree
+	collections map[string]*Collection
+}
+
+type Collection struct {
+	db         *Database
+	name       string
+	primaryKey []string
+}
+
+type Index struct {
+	collection *Collection
+	fields     []string
+}
+
+type Record struct {
+	Id   []byte
+	Data ipld.Node
+}
+
+type Query struct {
+	Equal map[string]ipld.Node
+}
+
+var NULL_BYTE = []byte("\x00")
+var FULL_BYTE = []byte("\xFF")
+var DATA_PREFIX = []byte("\x00d")
+var INDEX_PREFIX = []byte("\x00i")
+var DB_METADATA_KEY = []byte("\xFF\x00")
+var DB_VERSION = int64(1)
+var INDEX_VERSION_1 = int64(1)
+var VERSION_KEY = "version"
+
+func NewDatabaseFromBlockStore(ctx context.Context, blockStore blockstore.Blockstore) (*Database, error) {
+	nodeStore, err := tree.NewBlockNodeStore(blockStore, &tree.StoreConfig{CacheSize: 1 << 10})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// We're going to construct a fresh ProllyTree as our root to apply changes on
+	chunkConfig := tree.DefaultChunkConfig()
+	framework, err := tree.NewFramework(ctx, nodeStore, chunkConfig, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	collections := map[string]*Collection{}
+
+	// TODO: Document this with an IPLD Schema
+	metadata, err := qp.BuildMap(basicnode.Prototype.Any, -1, func(am datamodel.MapAssembler) {
+		qp.MapEntry(am, VERSION_KEY, qp.Int(DB_VERSION))
+		qp.MapEntry(am, "format", qp.String("database"))
+	})
+
+	// Initialize the tree with metadata saying it's an indexed tryee
+	framework.Append(ctx, DB_METADATA_KEY, metadata)
+
+	tree, rootCid, err := framework.BuildTree(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	db := &Database{
+		&blockStore,
+		nodeStore,
+		rootCid,
+		tree,
+		collections,
+	}
+
+	return db, nil
+}
+
+func NewMemoryDatabase() (*Database, error) {
+	ctx := context.Background()
+	blockStore := blockstore.NewBlockstore(datastore.NewMapDatastore())
+
+	return NewDatabaseFromBlockStore(ctx, blockStore)
+}
+
+func (db *Database) Flush(ctx context.Context) error {
+	rootCid, err := db.tree.Rebuild(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	db.rootCid = rootCid
+
+	return nil
+}
+
+func (db Database) StartMutating(ctx context.Context) error {
+	return db.tree.Mutate()
+}
+
+func (db Database) ExportToFile(ctx context.Context, destination string) error {
```

Comment:
Sweet, ty. Is there a case where not using methods is preferable?

`db.tree` in this case is a pointer so it'll have it's properties updated regardless of whether db is passed by value.

But yeah I'll convert them all to pointers to be safe. :)

Reply:
They're all methods, but yes sometimes using methods on values everywhere is used. For example `time.Time` is all methods on values because it was designed with the idea of never modifying the internal struct values, keeping each time struct as an immutable sort of unit, as if it was a native type.